src/graph/transform/cpu/knn.cc

```diff
@@ -15,6 +20,221 @@ namespace dgl {
 namespace transform {
 namespace impl {
 
+// This value is directly from pynndescent
+static constexpr int NN_DESCENT_BLOCK_SIZE = 16384;
+
+/*!
+ * \brief Compute Euclidean distance between two vectors, return positive
+ *  infinite value if the intermediate distance is greater than the worst
+ *  distance.
+ */
+template <typename FloatType, typename IdType>
+FloatType EuclideanDistWithCheck(const FloatType* vec1, const FloatType* vec2, int64_t dim,
+                                 FloatType worst_dist = std::numeric_limits<FloatType>::max()) {
+  FloatType dist = 0;
+  bool early_stop = false;
+  IdType idx = 0;
+  while (idx < dim - 3) {
+    FloatType diff0 = vec1[idx] - vec2[idx];
+    FloatType diff1 = vec1[idx + 1] - vec2[idx + 1];
+    FloatType diff2 = vec1[idx + 2] - vec2[idx + 2];
+    FloatType diff3 = vec1[idx + 3] - vec2[idx + 3];
+    dist += diff0 * diff0 + diff1 * diff1 + diff2 * diff2 + diff3 * diff3;
+    idx += 4;
+    if (dist > worst_dist) {
+      early_stop = true;
+      idx = dim;
+      break;
+    }
+  }
+
+  // last 3 elements
+  while (idx < dim) {
+    FloatType diff = vec1[idx] - vec2[idx];
+    dist += diff * diff;
+    ++idx;
+    if (dist > worst_dist) {
+      early_stop = true;
+      break;
+    }
+  }
+
+  if (early_stop) {
+    return std::numeric_limits<FloatType>::max();
+  } else {
+    return dist;
+  }
+}
+
+/*! \brief Compute Euclidean distance between two vectors */
+template <typename FloatType, typename IdType>
+FloatType EuclideanDist(const FloatType* vec1, const FloatType* vec2, int64_t dim) {
+  FloatType dist = 0;
+  IdType idx = 0;
+  while (idx < dim - 3) {
+    FloatType diff0 = vec1[idx] - vec2[idx];
+    FloatType diff1 = vec1[idx + 1] - vec2[idx + 1];
+    FloatType diff2 = vec1[idx + 2] - vec2[idx + 2];
+    FloatType diff3 = vec1[idx + 3] - vec2[idx + 3];
+    dist += diff0 * diff0 + diff1 * diff1 + diff2 * diff2 + diff3 * diff3;
+    idx += 4;
+  }
+
+  // last 3 elements
+  while (idx < dim) {
+    FloatType diff = vec1[idx] - vec2[idx];
+    dist += diff * diff;
+    ++idx;
+  }
+
+  return dist;
+}
+
+/*! \brief Insert a new element into a heap */
+template <typename FloatType, typename IdType>
+void HeapInsert(IdType* out, FloatType* dist,
+                IdType new_id, FloatType new_dist,
+                int k, bool check_repeat = false) {
+  if (new_dist > dist[0]) return;
+
+  // check if we have it
+  if (check_repeat) {
+    for (IdType i = 0; i < k; ++i) {
+      if (out[i] == new_id) return;
+    }
+  }
+
+  IdType left_idx = 0, right_idx = 0, curr_idx = 0, swap_idx = 0;
+  dist[0] = new_dist;
+  out[0] = new_id;
+  while (true) {
+    left_idx = 2 * curr_idx + 1;
+    right_idx = left_idx + 1;
+    swap_idx = curr_idx;
+    if (left_idx < k && dist[left_idx] > dist[swap_idx]) {
+      swap_idx = left_idx;
+    }
+    if (right_idx < k && dist[right_idx] > dist[swap_idx]) {
```

Comment:
Seems if `dist[left_idx] > dist[swap_idx]` and `dist[right_idx] > dist[swap_idx]` are both true, `swap_idx` is always set to be `right_idx`. Can we put the `if` about `right_idx` before, such that `left_idx` is under `else if`?

Reply:
If `dist[left_idx] > dist[swap_idx]` and `dist[right_idx] > dist[swap_idx]` are both true, then `dist[right_idx] > dist[left_idx]` must hold since the `swap_idx` is the index with biggest **current** distance.

Reply:
Yes, then we avoid a `if` execution for `left_idx`. Can we put `if (right_idx ...` first?

Reply:
I think since we want the biggest one among three elements, we need at least two comparisons here.

Reply:
Oh OK, here dist is a heap, i though it is a sorted list.